kotlin/flatbuffers-kotlin/src/commonMain/kotlin/com/google/flatbuffers/kotlin/JSON.kt

```diff
@@ -0,0 +1,788 @@
+            makeError(data, "Expecting a key/value after ${lastTok.print()}", tok)
+          }
+          lastTok = tok
+          tok = skipWhitespace(data)
+        }
+        LETTER_DOUBLE_QUOTE -> {
+          val childKey = readString(data)
+          tok = skipWhitespace(data)
+          if (tok.toChar() != ':')
+            makeError(data, "Expecting ':' after object Key", tok)
+          parseValue(data, skipWhitespace(data), childKey)
+          lastTok = tok
+          tok = skipWhitespace(data)
+          if (tok != LETTER_CLOSE_OBJECT && tok != LETTER_COMMA)
+            makeError(
+              data,
+              "Expecting '${LETTER_CLOSE_OBJECT.print()}' or '${LETTER_COMMA.print()}' after a key/value",
+              tok
+            )
+        }
+        else -> makeError(data, "Expecting start of object key", tok)
+      }
+    }
+    makeError(data, "Unable to parse the object", tok)
+  }
+
+  private fun parseArray(data: ReadBuffer, key: String? = null): FlexBufferType {
+    val fPos = output.startVector()
+    var lastTok: Byte = LETTER_OPEN_ARRAY // arrays always start with '['
+    var tok: Byte = skipWhitespace(data)
+    var elementType = T_INVALID
+    var multiType = false
+    val limit = data.limit
+    while (readPos <= limit) {
+      when (tok) {
+        LETTER_CLOSE_ARRAY -> {
+          if (lastTok == LETTER_COMMA) {
+            makeError(data, "Expecting a value after ${LETTER_COMMA.print()}", tok)
+          }
+          return if (!multiType && elementType.isScalar()) {
+            output.endTypedVector(fPos, key)
+            elementType.toElementTypedVector()
+          } else {
+            output.endVector(key, fPos)
+            T_VECTOR
+          }
+        }
+        LETTER_COMMA, 0.toByte() -> {
+          if (lastTok == LETTER_OPEN_ARRAY || lastTok == LETTER_COMMA) {
+            makeError(data, "Expecting a value after ${lastTok.print()}", tok)
+          }
+          lastTok = tok
+          tok = skipWhitespace(data)
+        }
+        LETTER_EOF -> makeError(data, "Expecting value or ${LETTER_CLOSE_ARRAY.print()}", tok)
+        else -> {
+          lastTok = LETTER_EOF // Anything but '[' since we will be reading a value
+          val newType = parseValue(data, tok)
+
+          if (elementType == T_INVALID) {
+            elementType = newType
+          } else if (newType != elementType) {
+            multiType = true
+          }
+          tok = skipWhitespace(data)
+          if (tok != LETTER_CLOSE_ARRAY && tok != LETTER_COMMA)
+            makeError(
+              data,
+              "Expecting ${LETTER_CLOSE_ARRAY.print()} or ${LETTER_COMMA.print()} after a value",
+              tok
+            )
+        }
+      }
+    }
+    makeError(data, "Unable to parse the array", tok)
+  }
+
+  private fun parseNumber(data: ReadBuffer, array: ByteArray, first: Byte, key: String?): FlexBufferType {
+    val ary = array
```

Comment:
I presume there's no code in Kotlin that does already what we want here? This kind of number parsing seems pretty common..

Reply:
There is, but it slow. We actually fall back to it if the number is not trivial x**200, etc.

That is why the number parsing is much faster than the other libs.